packages/firestore/src/core/query.ts

```diff
@@ -732,73 +732,80 @@ export const enum Direction {
-      "Bound has more components than query's orderBy"
-    );
-    let comparison = 0;
-    for (let i = 0; i < this.position.length; i++) {
-      const orderByComponent = orderBy[i];
-      const component = this.position[i];
-      if (orderByComponent.field.isKeyField()) {
-        debugAssert(
-          isReferenceValue(component),
-          'Bound has a non-key value where the key path is being used.'
-        );
-        comparison = DocumentKey.comparator(
-          DocumentKey.fromName(component.referenceValue),
-          doc.key
-        );
-      } else {
-        const docValue = doc.field(orderByComponent.field);
-        debugAssert(
-          docValue !== null,
-          'Field should exist since document matched the orderBy already.'
-        );
-        comparison = valueCompare(component, docValue);
-      }
-      if (orderByComponent.dir === Direction.DESCENDING) {
-        comparison = comparison * -1;
-      }
-      if (comparison !== 0) {
-        break;
-      }
+/**
+ * Returns true if a document sorts before a bound using the provided sort
+ * order.
+ */
+export function sortsBeforeDocument(
+  bound: Bound,
+  orderBy: OrderBy[],
+  doc: Document
+): boolean {
+  debugAssert(
+    bound.position.length <= orderBy.length,
+    "Bound has more components than query's orderBy"
+  );
+  let comparison = 0;
+  for (let i = 0; i < bound.position.length; i++) {
+    const orderByComponent = orderBy[i];
+    const component = bound.position[i];
+    if (orderByComponent.field.isKeyField()) {
+      debugAssert(
+        isReferenceValue(component),
+        'Bound has a non-key value where the key path is being used.'
+      );
+      comparison = DocumentKey.comparator(
+        DocumentKey.fromName(component.referenceValue),
+        doc.key
+      );
+    } else {
+      const docValue = doc.field(orderByComponent.field);
+      debugAssert(
+        docValue !== null,
+        'Field should exist since document matched the orderBy already.'
+      );
+      comparison = valueCompare(component, docValue);
+    }
+    if (orderByComponent.dir === Direction.DESCENDING) {
+      comparison = comparison * -1;
+    }
+    if (comparison !== 0) {
+      break;
     }
-    return this.before ? comparison <= 0 : comparison < 0;
   }
+  return bound.before ? comparison <= 0 : comparison < 0;
+}
 
-  isEqual(other: Bound | null): boolean {
-    if (other === null) {
-      return false;
-    }
-    if (
-      this.before !== other.before ||
-      this.position.length !== other.position.length
-    ) {
+export function boundEquals(left: Bound | null, right: Bound | null): boolean {
+  if (left === null) {
+    return right === null;
+  } else if (right === null) {
+    return false;
+  }
+
+  if (
+    left.before !== right.before ||
+    left.position.length !== right.position.length
+  ) {
+    return false;
+  }
+  for (let i = 0; i < left.position.length; i++) {
+    const thisPosition = left.position[i];
```

Comment:
nit: Consider renaming `thisPosition` to `leftPosition` and `otherPosition` to `rightPosition`.

Reply:
Done